### DISCUSSION TITLE
"No Functor ever provided with name 'dh_dt' under 'T_time' FVKernel" even though I listed the parameter

### DISCUSSION THREAD

**Discussion by waaizhaq** [idaholab/moose#29550] [Q&A Modules: Navier-Stokes]
### Check these boxes if you have followed the posting rules.

- [X] Q&A Navier Stokes is the most appropriate category for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (avoid screenshots if possible, triple back quotes before/after pasted text, etc)

### Question

Hello,

I am trying to conduct a simple natural convection problem in MOOSE, as shown below.

![Screenshot 2024-12-08 172412](https://github.com/user-attachments/assets/a9d10042-69a9-44e3-ba45-a07e7f3ab0fd)

My mesh only consists of the natural circulation flow, so the width of the wall is neglected. I just set a left boundary condition of Ts, and made my mesh thickness 1 meter.

I am very new to MOOSE so I am still learning how to use it. I've been following a lot of examples posted online in the framework to create my input file, so do let me know if something seems out of place.

But an error I've been running into is the 'dh_dt' not being provided even though I listed it under the INSFVEnergyTImeDerivative kernel. Is there something else that I'm missing?

I also tried running this problem as a steady state by commenting out the time-dependent kernels and changing the executioner type to steady, but then I got an error about the solve not converging.

By the way, to run this input I copied the navier stokes module into my own separate directory and compiled it, following the instructions that were in the framework under 'Pre-Built MOOSE.'

### Additional information

Mesh size and type: rectangular mesh, 5 by 5
 Discretization (finite element CG/DG, finite volume, etc): finite volume
 Models (turbulence, porous media, etc): natural convection
 Base input you started from: [nat_conv_wall.txt](https://github.com/user-attachments/files/18142394/nat_conv_wall.txt)

**Comment by GiudGiud** [comment]
Hello

This functor has to be defined by a FunctorMaterial. Here it should be this one:
https://mooseframework.inl.gov/source/functormaterials/INSFVEnthalpyFunctorMaterial.html

**Reply by waaizhaq** [reply]
Thank you! I'll add this to my input and see if I run into any other issues.

**Reply by waaizhaq** [reply]
So I've gotten past the FunctorMaterial issue, but I'm still running into convergence issues. I'm getting:

Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                   PC failed due to FACTOR_NUMERIC_ZEROPIVOT
  Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0

And I was looking at a similar discussion post that you answered a while back, mentioning to get around the FACTOR_NUMERIC_ZEROPIVOT, we can add a nonzero quantity to the diagonal, using the -pc_factor_shift NONZERO petsc options/values. I have that in my input but I'm still getting the same issue.

**Reply by GiudGiud** [reply]
should be `-pc_factor_shift_type` iirc

**Reply by waaizhaq** [reply]
That fixed it, thank you! I'm unfamiliar with the way MOOSE solves its simulations. Do you have any resources I can read into to understand the numerical methods of the different solvers?

**Reply by GiudGiud** [reply]
You ll have to look to the PETSc documentation for how the nonlinear system is solved. MOOSE does the assembly then calls petsc

**Reply by waaizhaq** [reply]
If I wanted to run the simulation as steady-state, which parameters would I change? I've commented out the time-dependent kernels and also the TimeStepper block under the Executioner block, then changed Transient to Steady, but then I run into another DIVERGED_DTOL convergence issue, I also added automatic_scaling in my executioner block.

Framework Information:
MOOSE Version:           git commit 2bd11a08a7 on 2024-10-06
LibMesh Version:
PETSc Version:           3.21.5
SLEPc Version:           3.21.1
Current Time:            Sun Dec 15 19:53:48 2024
Executable Timestamp:    Sun Oct  6 04:10:42 2024

Checkpoint:
  Wall Time Interval:      Every 3600 s
  User Checkpoint:         Disabled
  # Checkpoints Kept:      2
  Execute On:              TIMESTEP_END

Parallelism:
  Num Processors:          1
  Num Threads:             1

Mesh:
  Parallel Type:           replicated
  Mesh Dimension:          2
  Spatial Dimension:       2
  Nodes:                   36
  Elems:                   25
  Num Subdomains:          1

Nonlinear System:
  Num DOFs:                100
  Num Local DOFs:          100
  Variables:               { "vel_x" "vel_y" "pressure" "T" }
  Finite Element Types:    "MONOMIAL"
  Approximation Orders:    "CONSTANT"

Execution Information:
  Executioner:             Steady
  Solver Mode:             NEWTON
  PETSc Preconditioner:    lu
  MOOSE Preconditioner:    SMP (auto)


Postprocessor Values:
+----------------+----------------+
| time           | ave_T          |
+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |
+----------------+----------------+


Performing automatic scaling calculation

 0 Nonlinear |R| = 8.303678e+03
      0 Linear |R| = 8.303678e+03
      1 Linear |R| = 1.000718e-10
 1 Nonlinear |R| = 3.338182e+08
      0 Linear |R| = 3.338182e+08
      1 Linear |R| = 8.988101e-04
 2 Nonlinear |R| = 9.164821e+07
      0 Linear |R| = 9.164821e+07
      1 Linear |R| = 6.644202e-08
 3 Nonlinear |R| = 2.058075e+07
      0 Linear |R| = 2.058075e+07
      1 Linear |R| = 1.044626e-08
 4 Nonlinear |R| = 5.263572e+06
      0 Linear |R| = 5.263572e+06
      1 Linear |R| = 4.649094e-09
 5 Nonlinear |R| = 1.930216e+06
      0 Linear |R| = 1.930216e+06
      1 Linear |R| = 2.490749e-09
 6 Nonlinear |R| = 2.454784e+06
      0 Linear |R| = 2.454784e+06
      1 Linear |R| = 4.081857e-08
 7 Nonlinear |R| = 1.965131e+09
      0 Linear |R| = 1.965131e+09
      1 Linear |R| = 1.977286e-06
 8 Nonlinear |R| = 5.999012e+05
      0 Linear |R| = 5.999012e+05
      1 Linear |R| = 1.087908e-06
 9 Nonlinear |R| = 6.960520e+05
      0 Linear |R| = 6.960520e+05
      1 Linear |R| = 1.075856e-08
10 Nonlinear |R| = 1.063886e+06
      0 Linear |R| = 1.063886e+06
      1 Linear |R| = 3.278389e-09
11 Nonlinear |R| = 5.422805e+08
      0 Linear |R| = 5.422805e+08
      1 Linear |R| = 4.369014e-07
12 Nonlinear |R| = 7.352211e+06
      0 Linear |R| = 7.352211e+06
      1 Linear |R| = 3.356987e-06
13 Nonlinear |R| = 4.169012e+09
      0 Linear |R| = 4.169012e+09
      1 Linear |R| = 1.430620e-06
14 Nonlinear |R| = 9.723472e+09
      0 Linear |R| = 9.723472e+09
      1 Linear |R| = 1.466204e-03
15 Nonlinear |R| = 7.599592e+11
      0 Linear |R| = 7.599592e+11
      1 Linear |R| = 6.276364e-03
16 Nonlinear |R| = 1.370462e+13
      0 Linear |R| = 1.370462e+13
      1 Linear |R| = 5.705603e+00
17 Nonlinear |R| = 1.451590e+16
  Nonlinear solve did not converge due to DIVERGED_DTOL iterations 17
 Solve Did NOT Converge!
Aborting as solve did not converge

**Reply by GiudGiud** [reply]
For steady state (without using a relaxation transient) you need to work on the initialization. The initial condition needs to fall in the region of attraction of the solution, which can be hard to do

It's generally easier to use a relaxation-to-steady state transient

**Reply by waaizhaq** [reply]
I understand, I think that's what I have for my transient case, where it runs the simulation until it reaches steady state.
